### DISCUSSION TITLE
Metals diagnostics only providing syntax errors, no compile errors

### DISCUSSION THREAD

**Discussion by simonmclean-ovo** [scalameta/metals#4272] [Q&A] [answered]
On an M1 Macbook Pro. Here's the setup I've done so far:

1. `brew install coursier/formulas/coursier && cs setup`
3. I then installed JVM 17 with `cs java --jvm 17 --version` and `cs java --jvm 17 --setup`
5. Opened a Scala `2.13.8` project in vscode, and installed the Metals extension

**Problem**: I can compile the code with `sbt compile` just fine. If I  introduce a compile error (like delete a random import or a function param), then `sbt compile` will output the expected errors. However Metals will not detect the compile errors, and will only detect syntax errors.

Here's the output of Doctor

```
Build server currently being used is Bloop v1.5.3.

Metals Java: 17.0.4 from Eclipse Adoptium located at /Users/simon.mclean/Library/Caches/Coursier/arc/https/github.com/adoptium/temurin17-binaries/releases/download/jdk-17.0.4%252B8/OpenJDK17U-jdk_x64_mac_hotspot_17.0.4_8.tar.gz/jdk-17.0.4+8/Contents/Home

Metals Server version: 0.11.8

[table redacted]

Compilation status:

✅ - code is compiling
❌ - code isn't compiling, open problems tab to see detailed compilation errors
Diagnostics:

✅ - diagnostics correctly being reported by the build server
⚠️ - only syntactic errors being reported
Interactive features (completions, hover):

✅ - supported Scala version
Semanticdb features (references, renames, go to implementation):

✅ - build tool automatically creating needed semanticdb files
Debugging (run/test, breakpoints, evaluation):

✅ - users can run or test their code with debugging capabilities
❌ - the tool does not support debugging in this target
Java Support:

✅ - working non-interactive features (references, rename etc.)
```

If I open vscode's "PROBLEMS" tab, I can see this:

```
value ofInstant is not a member of object java.time.LocalDate (bloop)
```

`ofInstant` definitely is a member of `LocalDate` in Java 17, but it isn't in much older versions. So my suspicion is that somehow bloop is not using the same Java version which my `JAVA_HOME` variable is pointing to. But my understanding is the Metals is supposed to forward the `JAVA_HOME` value it's using to Bloop. What's more I've checked `$HOME/.bloop/.bloop.json`, and it seems correct, so I've no idea how this is happening...

**Comment by tanishiking** [comment, marked as answer]
@simonmclean-ovo Thank you for reporting.
Seems like you already checked, but
- Does HOME/.bloop/bloop.json's jqvaHome field point to Java17?
- Did you restart bloop after updating the global bloop file?

If both true, I also have no idea what's going on...

---

> However Metals will not detect the compile errors, and will only detect syntax errors.

Is this another issue than the Java17 stuff? LocalDate error seems not a syntax error, it's a semantic error.

**Reply by simonmclean-ovo** [reply]
@tanishiking

> Did you restart bloop after updating the global bloop file

Not explicitly. I just assumed that if it's no sbt session running, that bloop wouldn't be running either. Is this not the case?

> LocalDate error seems not a syntax error, it's a semantic error

True, but it does seem to be preventing Metals from reporting any other other semantic errors

**Reply by tgodzik** [reply]
Bloop is running underneath to provide compilation. It should actually say which version it is using in the logs, something along the lines of:
```
2022.08.15 18:15:26 INFO  Bloop uses /usr/lib/jvm/openjdk17 defined at /home/tgodzik/.bloop/bloop.json
2022.08.15 18:15:26 INFO  Bloop currently uses settings: -Xmx3G
```
do you see it?

**Reply by tanishiking** [reply]
> I just assumed that if it's no sbt session running, that bloop wouldn't be running either. Is this not the case?

Not really. Metals use `bloop` under the hood to compile the project and emit diagnoses.
So, as long as Metals running, `bloop` is there. What happens if you restart bloop? by `Metals: restart build server`.
 
> it does seem to be preventing Metals from reporting any other other semantic errors

Hmm, what else errors should be there?

**Reply by simonmclean-ovo** [reply]
This seems to be fixed now. I just restarted my machine lol. I think it probably was bloop using the outdated `jvmHome` value, as you suggested @tanishiking 🤦

**Reply by tgodzik** [reply]
Next time you should be able to run the `Restart build server` command which should pick up the new java home. We also show a notification, but that's sometimes easy to miss.
